mountpoint-s3-client/src/mock_client.rs

```diff
@@ -233,8 +233,9 @@ impl ObjectClient for MockClient {
         bucket: &str,
         key: &str,
         range: Option<Range<u64>>,
+        if_match: Option<String>,
```

Comment:
I think we should define an `ETag` struct somewhere rather than just always handling them as strings. Just:
```rust
pub struct ETag(String);
```
That makes the API a lot more obvious, and also prevents them from being invented "from thin air"--the only way to get an ETag is through `S3CrtClient` methods that return them.

Reply:
I am not really sure how to use this ETag struct. Because String and ETag struct are different and they cant be used interchangeably. Maybe, I am not understanding its use case correctly.

Reply:
It's two principles we're trying to enforce:
1. By making the API take an `ETag` as an argument we make clear that it's not an arbitrary string input, but a specific type of input that's required.
2. The `ETag` struct can use [visibility/privacy](https://doc.rust-lang.org/reference/visibility-and-privacy.html) to prevent clients from "faking" an ETag by making the fields not visible.

You would probably write it like this, in `object_client.rs`:
```rust
pub struct ETag(pub(crate) String);
```
or like this:
```rust
pub struct ETag {
    pub(crate) etag: String.
}
```
(They're equivalent, just a taste question about whether you prefer the field having a name or not).

This is saying that any code in the `mountpoint-s3-client` struct can see inside the `ETag` struct to use the actual String. But code *outside* that struct can't see the field (`etag.0` or `etag.etag` won't compile), and also can't create a new ETag by writing `ETag("foo".into())` or `ETag { etag: "foo".into() }`.

What this gets us is a guarantee that `ETag`s  are always "real", in the sense that they were only created by the `mountpoint-s3-client` crate. A consumer can't, for example, accidentally create an ETag out of the wrong string, like this dumb example:
```rust
let object = client.head_object(...).await.unwrap();
let etag: &str = object.bucket; // oops, typo!
let data = client.get_object(..., etag).await.unwrap();
```
With the `ETag` struct, this silly typo won't compile.

Reply:
Thanks a lot. Understood the use case.